htlcswitch/switch.go

```diff
@@ -63,6 +64,50 @@ var (
 	ErrUnreadableFailureMessage = errors.New("unreadable failure message")
 )
 
+// forwardResolver implements the htlcinterceptor.Interceptor interface.
+// It is passed from the switch to external interceptors that are interested
+// in holding forwards and resolve them manually.
+type forwardResolver struct {
+	packet     *htlcPacket
+	htlcSwitch *Switch
+}
+
+// Resume resumes the default behavior as if the packet was not intercepted.
+func (h *forwardResolver) Resume() error {
+	return h.htlcSwitch.handlePacketForward(h.packet)
+}
+
+// Fail forward a failed packet to the switch.
+func (h *forwardResolver) Fail() error {
+	return h.resolve(&lnwire.UpdateFailHTLC{})
+}
+
+// Settle forwards a settled packet to the switch.
+func (h *forwardResolver) Settle(preimage lntypes.Preimage) error {
+	return h.resolve(&lnwire.UpdateFulfillHTLC{
+		PaymentPreimage: preimage,
+	})
+}
+
+// resolve is used for both Settle and Fail and forwards the message to the
+// switch. When creating the forwarded packet this method also populates the
+// circuitInKey field which enables the switch to find the circuit before
+// forwarding the result to the incoming link, this is needed because
+// intercepted packets don't have an opened circuit in the switch as they were
+// intercepted before they made it to the outgoing link.
+func (h *forwardResolver) resolve(message lnwire.Message) error {
+	circuitInKey := h.packet.inKey()
+	pkt := &htlcPacket{
+		outgoingChanID: h.packet.outgoingChanID,
+		outgoingHTLCID: h.packet.outgoingHTLCID,
+		isResolution:   true,
+		circuit:        h.packet.circuit,
+		circuitInKey:   &circuitInKey,
+		htlc:           message,
+	}
+	return h.htlcSwitch.handlePacketForward(pkt)
```

Comment:
I _think_ this should either be calling `ForwardPackets`, `routeAsync` or `route`. As far as I can tell, calls to `handlePacketForward` always happen in the context of `hltcForwarder` go routine to keep the switch "sync". Calling it here directly might be messing up some assumptions made down the line. The 3 methods suggested before make sure the message ends up in `hltcForwarder`.

Reply:
Good one! Used route.